src/types/datetime.rs

```diff
@@ -26,8 +25,62 @@ use crate::ffi::{
 use crate::types::PyTuple;
 use crate::{AsPyPointer, PyAny, PyObject, Python, ToPyObject};
 use std::os::raw::c_int;
-#[cfg(not(PyPy))]
-use std::ptr;
+
+fn ensure_datetime_api(_py: Python) -> &'static PyDateTime_CAPI {
+    unsafe {
+        if pyo3_ffi::PyDateTimeAPI.get().is_null() {
+            PyDateTime_IMPORT()
+        }
+
+        &*pyo3_ffi::PyDateTimeAPI.get()
+    }
+}
+
+// Type Check macros
+//
+// These are bindings around the C API typecheck macros, all of them return
+// `1` if True and `0` if False. In all type check macros, the argument (`op`)
+// must not be `NULL`. The reexported implementations here call ensure_datetime_api is initalized
+//
+// # Safety
+//
+// These functions must only be called when the GIL is held!
+
+macro_rules! reexport_with_autoinit {
```

Comment:
as these aren't re-exports any more, just wrappers, might be nice to give the macro a different name.